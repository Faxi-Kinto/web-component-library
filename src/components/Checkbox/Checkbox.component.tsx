import React, { useState, ChangeEvent, useMemo } from 'react';
import * as Styled from './Checkbox.styles';

import Text from '../Text';

/**
 * @name Checkbox
 *
 * [Insert Component Description]
 *
 * @returns {JSX}
 */

type LabelPosition = 'left' | 'right';

export type CheckboxProps = {
  error?: boolean;
  name?: string;
  checked?: boolean;
  label?: React.ReactNode;
  labelPosition?: LabelPosition;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
};

const Checkbox: React.FC<CheckboxProps> = (
  props: CheckboxProps
): JSX.Element => {
  const { checked, onChange, name, label, error, labelPosition } = props;
  const [stateChecked, setChecked] = useState(checked);

  function handleInputOnChange(event: ChangeEvent<HTMLInputElement>): void {
    if (onChange) onChange(event);
    else setChecked(event.target.checked);
  }

  const labelEl = useMemo(() => {
    if (!label) return null;

    if (typeof label === 'string') {
      return <Text.Body>{label}</Text.Body>;
    }

    return label;
  }, [label]);

  return (
    <Styled.Container
      className={`${label ? 'label--has-label-text ' : ''}${
        checked || stateChecked ? 'checked ' : ''
      }${error ? 'label--has-error ' : ''}${
        label && labelPosition === 'left' ? 'label--left' : ''
      }`}
    >
      <input
        type="checkbox"
        name={name}
        checked={checked || stateChecked}
        onChange={handleInputOnChange}
      />
      {labelEl}
    </Styled.Container>
  );
};

Checkbox.defaultProps = {
  checked: false,
  label: '',
};

export default Checkbox;
