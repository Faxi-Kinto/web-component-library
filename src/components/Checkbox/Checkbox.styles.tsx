import styled from 'styled-components';
import { theme } from '../../config';
import { size, positionAbsolute, padding } from '../../styles/basics';
import { flexRow } from '../../styles/flex';
import { pxToRem } from '../../styles/basics';

const defaultSize = pxToRem('20px');

export const Container = styled.label`
  position: relative;
  cursor: pointer;
  min-height: ${defaultSize};
  padding-left: ${pxToRem('20px')};
  ${flexRow('initial', 'center')};

  > input {
    display: none;
  }

  &:before {
    content: '';
    display: inline-block;
    box-sizing: border-box;
    border: ${theme.pallet.ACCENT_1} solid ${pxToRem('1px')};
    ${positionAbsolute('0', '', '', '0')};
    ${size(defaultSize)};
  }

  > span {
    line-height: ${defaultSize};
  }

  &.checked {
    &:after {
      background: url('assets/icons/checkmark.svg') 100% no-repeat;
      content: '';
      ${size(pxToRem('13px'), pxToRem('9px'))};
      ${positionAbsolute(pxToRem('5.5px'), '', '', pxToRem('3px'))};
    }
  }

  &.label--has-label-text {
    padding-left: ${pxToRem('28px')};
  }

  &.label--has-error {
    > input {
      color: ${theme.pallet.ERROR};
    }

    &:before {
      border: ${theme.pallet.ERROR} solid ${pxToRem('1px')};
    }
  }

  &.label--left {
    ${padding('0', pxToRem('28px'), '0', '0')};

    &:before {
      ${positionAbsolute('0', '0', '', 'auto')};
    }

    &.checked {
      &:after {
        ${positionAbsolute(pxToRem('5.5px'), pxToRem('4px'), '', 'auto')}
      }
    }
  }
`;
