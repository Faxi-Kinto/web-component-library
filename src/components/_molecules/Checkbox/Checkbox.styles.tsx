import styled from 'styled-components';
import {
  size,
  positionAbsolute,
  flexRow,
  pxToRem,
} from '@faxi/web-css-utilities';

type CheckboxStylingProps = {
  borderColor?: string;
  errorColor?: string;
  size?: string;
};

const defaultSize = pxToRem('20px');

export const Container = styled.label`
  position: relative;
  cursor: pointer;
  display: flex;
  ${size('100%', '')};
  min-height: ${(props: CheckboxStylingProps) =>
    props.size ? props.size : defaultSize};

  > span {
    line-height: ${(props: CheckboxStylingProps) =>
      props.size ? props.size : defaultSize};
  }

  &:before {
    content: '';
    display: inline-block;
    box-sizing: border-box;
    border: ${(props: CheckboxStylingProps) =>
      `${props.borderColor} solid ${pxToRem('1px')}`};
    ${positionAbsolute('0', '', '', '0')};
    ${(props: CheckboxStylingProps) =>
      props.size
        ? `
      ${size(props.size)};
      `
        : `
      ${size(defaultSize)};
      `};
  }

  &.checkbox {
    &--has-label {
      & > :last-child {
        margin-left: ${(props: CheckboxStylingProps) =>
          props.size ? props.size : defaultSize};
        padding-left: ${pxToRem('5px')};
      }
    }

    &--has-error {
      color: ${(props: CheckboxStylingProps) => `${props.errorColor}`};
      &:before {
        border: ${(props: CheckboxStylingProps) => `${props.errorColor}`} solid
          ${pxToRem('1px')};
      }
    }

    &--left {
      & > :last-child {
        margin-left: unset;
        padding-left: unset;
        margin-right: ${(props: CheckboxStylingProps) =>
          props.size ? props.size : defaultSize};
        padding-right: ${pxToRem('5px')};
      }

      &:before {
        ${positionAbsolute('0', '0', '', 'auto')};
      }

      .checkbox {
        &__input-wrapper {
          ${positionAbsolute('0', '0', '', 'auto')};
        }
      }
    }
  }

  .checkbox {
    &__input-wrapper {
      position: absolute;
      ${(props: CheckboxStylingProps) =>
        props.size
          ? `
      ${size(props.size)};
      `
          : `
      ${size(defaultSize)};
      `};
      ${flexRow('center', 'center')};

      &__input {
        display: none;
      }
    }
  }
`;
