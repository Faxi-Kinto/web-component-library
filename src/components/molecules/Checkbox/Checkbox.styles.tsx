import styled from 'styled-components';
import {
  size,
  positionAbsolute,
  flexRow,
  pxToRem,
} from '@faxi/web-css-utilities';

type CheckboxStylingProps = {
  borderColor: string;
  errorColor?: string;
};

const defaultSize = pxToRem('26px');

export const Container = styled.label`
  position: relative;
  cursor: pointer;
  display: flex;
  ${size('fit-content', defaultSize)};

  > span {
    line-height: ${defaultSize};
  }

  &:before {
    content: '';
    display: inline-block;
    box-sizing: border-box;
    border: ${(props: CheckboxStylingProps) =>
      `${props.borderColor} solid ${pxToRem('1px')}`};
    ${positionAbsolute('0', '', '', '0')};
    ${size(defaultSize)};
  }

  &.checkbox {
    &--has-label {
      & > :last-child {
        margin-left: ${defaultSize};
        padding-left: ${pxToRem('5px')};
      }
    }

    &--has-error {
      color: ${(props: CheckboxStylingProps) => `${props.errorColor}`};
      &:before {
        border: ${(props: CheckboxStylingProps) => `${props.errorColor}`} solid
          ${pxToRem('1px')};
      }
    }

    &--left {
      & > :last-child {
        margin-left: unset;
        padding-left: unset;
        margin-right: ${defaultSize};
        padding-right: ${pxToRem('5px')};
      }

      &:before {
        ${positionAbsolute('0', '0', '', 'auto')};
      }

      .checkbox {
        &__input-wrapper {
          ${positionAbsolute('0', '0', '', 'auto')};
        }
      }
    }
  }

  .checkbox {
    &__input-wrapper {
      position: absolute;
      ${size(defaultSize, '100%')}
      ${flexRow('center', 'center')};

      &__input {
        display: none;
      }
    }
  }
`;
