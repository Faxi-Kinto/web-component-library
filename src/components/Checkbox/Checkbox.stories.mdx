import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";

import Checkbox from "./";
import Text from '../../atoms/Text'

<Meta title="Core|Molecules/Checkbox" />

# Checkbox Component

Simple checkbox component. It can be simple, or with label

## Implementation

#### Default
<Preview>
  <Story name="default">
    <Checkbox />
  </Story>
</Preview>


#### With Label
<Preview>
  <Story name="With label">
    <Checkbox label="Click me!" />
  </Story>
</Preview>

#### With link
<Preview>
  <Story name="With link">
    <Checkbox label={
      <Text.Body>I accept&nbsp;
        <Text.Link href="https://google.com">terms</Text.Link>
      </Text.Body>} />
  </Story>
</Preview>

#### With error
<Preview>
  <Story name="With error">
    <Checkbox error label={'This won\'t work'} />
  </Story>
</Preview>


#### Label left
<Preview>
  <Story name="Label left">
    <Checkbox label={'This is left'} labelPosition={'left'} />
  </Story>
</Preview>

## Props



<Props of={Checkbox} />


## Notes
- onChange prop passes Event Object through callback. Reason for not passing only check 
property is having full control over the event **_e.g._** `event.preventDefault() || event.stopPropagation()`