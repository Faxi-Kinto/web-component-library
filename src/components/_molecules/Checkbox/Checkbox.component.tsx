import React, {
  useState,
  ChangeEvent,
  useMemo,
  ReactNode,
  Fragment,
} from 'react';
import * as Styled from './Checkbox.styles';
import Text from '../../_atoms/Text';
import classNames from 'classnames';

/**
 * @name Checkbox
 *
 * [Insert Component Description]
 *
 * @returns {JSX}
 */

type LabelPosition = 'left' | 'right';

export type CheckboxProps = {
  error?: boolean;
  name?: string;
  value?: boolean;
  label?: React.ReactNode;
  labelPosition?: LabelPosition;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  borderColor?: string;
  errorColor?: string;
  icon: JSX.Element;
  size?: string;
  iconClassName?: string;
};

const Checkbox: React.FC<CheckboxProps> = (
  props: CheckboxProps
): JSX.Element => {
  const {
    value,
    onChange,
    name,
    label,
    error,
    labelPosition,
    errorColor = 'red',
    borderColor,
    icon,
    size,
    iconClassName,
  } = props;
  const [stateChecked, setChecked] = useState(Boolean(value));

  const handleInputOnChange = (event: ChangeEvent<HTMLInputElement>): void => {
    if (onChange) onChange(event);
    else setChecked(event.target.checked);
  };

  const labelEl = useMemo(() => {
    if (!label) return null;

    if (typeof label === 'string') {
      return <Text.Body>{label}</Text.Body>;
    }

    return label;
  }, [label]);

  return (
    <Styled.Container
      borderColor={borderColor}
      errorColor={errorColor}
      size={size}
      className={`checkbox${label ? ' checkbox--has-label' : ''}${
        value || stateChecked ? ' checkbox--checked' : ''
      }${error ? ' checkbox--has-error ' : ''}${
        label && labelPosition === 'left' ? ' checkbox--left' : ''
      }`}
    >
      <div className="checkbox__input-wrapper">
        <input
          className="checkbox__input-wrapper__input"
          type="checkbox"
          name={name}
          checked={Boolean(value || stateChecked)}
          value={`${Boolean(value || stateChecked)}`}
          onChange={handleInputOnChange}
        />
        {value || stateChecked ? (
          <Fragment>
            {icon &&
              React.cloneElement(icon, {
                className: classNames([
                  'checkbox__input-wrapper__input__icon',
                  iconClassName,
                ]),
              })}
          </Fragment>
        ) : null}
      </div>
      {labelEl}
    </Styled.Container>
  );
};

Checkbox.defaultProps = {
  label: '',
};

export default Checkbox;
