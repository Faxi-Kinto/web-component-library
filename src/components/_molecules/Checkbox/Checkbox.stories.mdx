import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";

import Checkbox from "./";
import Text from '../../_atoms/Text'

<Meta title="Core|Molecules/Checkbox" />

# Checkbox Component

You can provide your own checked state icon

## Implementation

#### Default
<Preview>
  <Story name="default">
    <Checkbox icon={
      <img style={{width: '70%'}} src="assets/icons/checkmark.svg" />
      } />
  </Story>
</Preview>


#### With Label
<Preview>
  <Story name="With label and size of 50px">
    <Checkbox label="Click me!" icon={
      <img style={{width: '70%'}} src="assets/icons/checkmark.svg" />
      }  size="50px" />
  </Story>
</Preview>

#### With link
<Preview>
  <Story name="With link">
    <Checkbox label={
      <Text.Body>I accept&nbsp;
        <Text.Link href="https://google.com">terms</Text.Link>
      </Text.Body>} icon={
      <img style={{width: '70%'}} src="assets/icons/checkmark.svg" />
      } />
  </Story>
</Preview>

#### With error
<Preview>
  <Story name="With error">
    <Checkbox error label={'This won\'t work'} icon={
      <img style={{width: '70%'}} src="assets/icons/checkmark.svg" />
      }  />
  </Story>
</Preview>


#### Label left
<Preview>
  <Story name="Label left">
    <Checkbox label={'This is left'} labelPosition={'left'} icon={
      <img style={{width: '70%'}} src="assets/icons/checkmark.svg" />
      }  />
  </Story>
</Preview>

#### Colored
<Preview>
  <Story name="Colored">
    <Checkbox icon={
      <img style={{width: '70%'}} src="assets/icons/checkmark.svg" />
      } borderColor={'green'}/>
  </Story>
</Preview>

## Props

<Props of={Checkbox} />

## Notes
- onChange prop passes Event Object through callback. Reason for not passing only check 
property is having full control over the event **_e.g._** `event.preventDefault() || event.stopPropagation()`