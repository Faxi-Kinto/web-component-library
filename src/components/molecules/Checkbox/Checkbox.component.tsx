import React, { useState, ChangeEvent, useMemo } from 'react';
import * as Styled from './Checkbox.styles';
import Text from '../../atoms/Text';
import Icon from '../../atoms/Icon';

/**
 * @name Checkbox
 *
 * [Insert Component Description]
 *
 * @returns {JSX}
 */

type LabelPosition = 'left' | 'right';

export type CheckboxProps = {
  error?: boolean;
  name?: string;
  value?: boolean;
  label?: React.ReactNode;
  labelPosition?: LabelPosition;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
};

const Checkbox: React.FC<CheckboxProps> = (
  props: CheckboxProps
): JSX.Element => {
  const { value, onChange, name, label, error, labelPosition } = props;
  const [stateChecked, setChecked] = useState(Boolean(value));

  const handleInputOnChange = (event: ChangeEvent<HTMLInputElement>): void => {
    if (onChange) onChange(event);
    else setChecked(event.target.checked);
  };

  const labelEl = useMemo(() => {
    if (!label) return null;

    if (typeof label === 'string') {
      return <Text.Body>{label}</Text.Body>;
    }

    return label;
  }, [label]);

  return (
    <Styled.Container
      className={`checkbox${label ? ' checkbox--has-label ' : ''}${
        value || stateChecked ? ' checkbox--checked ' : ''
      }${error ? ' checkbox--has-error ' : ''}${
        label && labelPosition === 'left' ? ' checkbox--left' : ''
      }`}
    >
      <div className="checkbox__input-wrapper">
        <input
          className="checkbox__input-wrapper__input"
          type="checkbox"
          name={name}
          checked={Boolean(value || stateChecked)}
          value={`${Boolean(value || stateChecked)}`}
          onChange={handleInputOnChange}
        />
        {value || stateChecked ? <Icon name="check" /> : null}
      </div>
      {labelEl}
    </Styled.Container>
  );
};

Checkbox.defaultProps = {
  label: '',
};

export default Checkbox;
