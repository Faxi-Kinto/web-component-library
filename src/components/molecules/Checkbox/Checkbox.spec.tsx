import React from 'react';
import renderer from 'react-test-renderer';
import Checkbox from './';

test('Checkbox renders correctly', () => {
  const component = renderer.create(<Checkbox />);

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
