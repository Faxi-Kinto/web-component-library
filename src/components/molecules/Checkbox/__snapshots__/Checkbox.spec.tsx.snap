// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders correctly 1`] = `
.c0 {
  position: relative;
  cursor: pointer;
  min-height: 1.25rem;
  padding-left: 1.25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: initial;
  -webkit-justify-content: initial;
  -ms-flex-pack: initial;
  justify-content: initial;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 > input {
  display: none;
}

.c0:before {
  content: '';
  display: inline-block;
  box-sizing: border-box;
  border: #4E606A solid 0.0625rem;
  position: absolute;
  top: 0;
  left: 0;
  width: 1.25rem;
  height: 1.25rem;
}

.c0 > span {
  line-height: 1.25rem;
}

.c0.checked:after {
  background: url('assets/icons/checkmark.svg') 100% no-repeat;
  content: '';
  width: 0.8125rem;
  height: 0.5625rem;
  position: absolute;
  top: 0.34375rem;
  left: 0.1875rem;
}

.c0.label--has-label-text {
  padding-left: 1.75rem;
}

.c0.label--has-error > input {
  color: #EB4820;
}

.c0.label--has-error:before {
  border: #EB4820 solid 0.0625rem;
}

.c0.label--left {
  padding: 0 1.75rem 0 0;
}

.c0.label--left:before {
  position: absolute;
  top: 0;
  right: 0;
  left: auto;
}

.c0.label--left.checked:after {
  position: absolute;
  top: 0.34375rem;
  right: 0.25rem;
  left: auto;
}

<label
  className="c0"
>
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />
</label>
`;
